docs/training/apis-howto/api-core-mnist-ug.rst

```diff
@@ -0,0 +1,573 @@
+   core_context.train.report_validation_metrics(
+       steps_completed=steps_completed,
+       metrics={"test_loss": test_loss},
+   )
+
+.. tip::
+
+   Avoiding a Duplicate Key Error
+
+   It is best to stitch metrics together in one continuous graph as your experiment progresses. To
+   accomplish this, calculate the index used for ``steps_completed`` appropriately. The appropriate
+   calculation avoids re-writing metrics with the same index each time a new epoch begins and avoids
+   a duplicate key error.
+
+Step 2.3: Modify the Test Method
+================================
+
+Modify the ``test()`` function header to include ``args`` and other elements you’ll need during the
+evaluation loop. In addition, pass the newly created ``core_context`` into both ``train()`` and
+``test()``.
+
+.. code::
+
+   # NEW: Pass core_context into train() and test().
+   train(args, model, device, train_loader, optimizer, epoch, core_context)
+
+   # NEW: Pass args, test_loader, epoch, and steps_completed into test().
+   test(args, model, device, test_loader, epoch, core_context, steps_completed=steps_completed)
+
+Create a ``steps_completed`` variable to plot metrics on a graph in the WebUI:
+
+.. note::
+
+   Refer to line 198 in ``model_def_metrics.py``
+
+.. code::
+
+   # NEW: Calculate steps_completed for plotting test metrics.
+   steps_completed = epoch * len(train_loader)
+
+Step 2.4: Run the Experiment
+============================
+
+Run the following command to run the experiment:
+
+Finally, run the following command to run the experiment:
+
+.. code::
+
+   det e create metrics.yaml .
+
+Open the Determined WebUI again and navigate to the **Overview** tab.
+
+The WebUI now displays metrics. In this step, you learned how to add a few new lines of code in
+order to report training and validation metrics to the Determined master. Next, we’ll modify our
+script to report checkpoints.
+
+***********************
+ Step 3: Checkpointing
+***********************
+
+Checkpointing periodically during training and reporting the checkpoints to the master gives us the
+ability to stop and restart training. In this section, we’ll modify our script for the purpose of
+checkpointing.
+
+..
+   note::Two Types of Continuations (Stopping and Restarting)
+
+   There are two types of continuations (methods for stopping and restarting
+   training): pausing and reactivating training using the WebUI or clicking
+   **Continue Trial** after the experiment completes.
+
+   Each type of continuation has its own behavior. While you always want to preserve
+   the value you are incrementing (the “model weight”), you do not always
+   want to preserve the batch index. When you pause and reactivate you want
+   training to continue from the same batch index. However, when starting a
+   fresh experiment, you want training to start with a fresh batch index.
+
+   To distinguish between the two types of continuations, you can save the
+   trial ID in the checkpoint.
```

Comment:
This comment is great and all, but the mnist example doesn't actually handle any of this nuance.

I think we should update the example and keep the note, rather than discard the note.

Reply:
note: pause and reactivate

# NEW: Detect when the experiment is paused by the WebUI.
        if core_context.preempt.should_preempt():
            return

Reply:
Pushed new changes to model_def_checkpoints.py, model_def_adaptive.py, and model_def_distributed.py to reflect this nuance. Tested model_def_checkpoints.py for both cases, waiting for experiments to complete to test stopping and continuing for adaptive search and distributed training. Meanwhile @rb-determined-ai let me know if there's anything that needs to be changed in the way I implemented it.

Reply:
user guide updated:

Optional: Enable Continuing the Trial
To enable continuing the trial after the experiment completes, save the trial ID. One way to do this is to load the checkpoint and save the checkpoint in a file in the checkpoint directory.

Open the checkpoint.pt file in binary mode and compare ckpt_trial_id with the current trial_id.

   with checkpoint_directory.joinpath("state").open("r") as f:
    epochs_completed, ckpt_trial_id = [int(field) for field in f.read().split(",")]

# If trial ID does not match our current trial ID, we'll ignore epochs
# completed and start training from epoch_idx = 0
if ckpt_trial_id != trial_id:
    epochs_completed = 0
Save the checkpoint in the checkpoint.pt file.

   # NEW: Save checkpoint.
checkpoint_metadata_dict = {"steps_completed": steps_completed}

# NEW: Here we are saving multiple files to our checkpoint directory.
# 1) a model state file and 2) a file includes information
# about the training loop state.
with core_context.checkpoint.store_path(checkpoint_metadata_dict) as (path, storage_id):
    torch.save(model.state_dict(), path / "checkpoint.pt")
    with path.joinpath("state").open("w") as f:
        f.write(f"{epochs_completed},{info.trial.trial_id}")